src/kernels/MIOpenCartesianProd.cpp

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename T>
+__device__ void cartesianProdForward(const T* input,
+                                     T* output_ws,
+                                     tensor_view_t<1> input_tv,
+                                     tensor_view_t<2> output_tv,
+                                     size_t stride,
+                                     size_t dim1_idx)
+{
+    size_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+    if(gid >= output_tv.size[0])
+        return;
+    output_ws[output_tv.size[0] * dim1_idx + gid] = input[(gid / stride) % input_tv.size[0]];
```

Comment:
May return invalid output if `input_tv.stride[0] != 1`
```suggestion
    output_ws[output_tv.size[0] * dim1_idx + gid] = input[input_tv.get_tensor_view_idx({(gid / stride) % input_tv.size[0]})];
```

Reply:
resolved